src/main/java/software/amazon/smithy/lsp/SmithyTextDocumentService.java

```diff
@@ -197,6 +225,97 @@ private String findToken(String path, Position p) throws IOException {
 
         }
 
+        return contents;
+    }
+
+    private Position blankPosition = new Position(-1, 0);
+    private Position documentStart = new Position(0, 0);
+
+    private TextEdit insertLine(String line, DocumentPreamble preamble) {
+        String trailingNewLine;
+        if (!preamble.isBlankSeparated())
+            trailingNewLine = "\n";
+        else
+            trailingNewLine = "";
+        // case 1 - there's no use block at all, so we need to insert the line directly
+        // under namespace
+        if (preamble.getUseBlockRange().getStart() == blankPosition) {
+            // case 1.a - there's no namespace - that means the document is invalid
+            // so we'll just insert the line at the beginning of the document
+            if (preamble.getNamespaceRange().getStart() == blankPosition) {
+                return new TextEdit(new Range(documentStart, documentStart), line + trailingNewLine);
+            } else {
+                Position namespaceEnd = preamble.getNamespaceRange().getEnd();
+                namespaceEnd.setCharacter(namespaceEnd.getCharacter() + 1);
+                return new TextEdit(new Range(namespaceEnd, namespaceEnd), "\n" + line + trailingNewLine);
+            }
+        } else {
+            Position useBlockEnd = preamble.getUseBlockRange().getEnd();
+            return new TextEdit(new Range(useBlockEnd, useBlockEnd), "\n" + line + trailingNewLine);
+        }
+    }
```

Comment:
After we've identified the preamble, it's easier to construct a valid text edit - only thing we need to think about is duplication of newlines after preamble.